compiler/rustc_hir/src/def.rs

```diff
@@ -440,6 +440,13 @@ impl DefKind {
             | DefKind::ExternCrate => false,
         }
     }
+
+    /// Returns `true` if `self` is a kind of definition that does not have its own
+    /// type-checking context, i.e. closure, coroutine or inline const.
+    #[inline]
+    pub fn is_typeck_child(self) -> bool {
+        matches!(self, DefKind::Closure | DefKind::InlineConst | DefKind::SyntheticCoroutineBody)
```

Comment:
`typeck_root_def_id == self.def_id` instead of matching on the `DefKind`.

This is brittle as it's quite easy to add a new `DefKind` which is also typechecked with their parent.

Reply:
This match cannot be avoided and I simply moved it. However at least I could make it exhaustive if you like.